### PR TITLE
run mypy in github actions, remove mypy from tests, use strict mode flags

### PR DESCRIPTION
**Pull request by DustinAlandzes**
removed mypy from the tests

I also removed test.sh, but I see how this is useful so would you like me to re-add it?

### CODE REVIEW

File: slippi/event.py [hunk 1]
```diff
@@ -260,21 +268,21 @@ class Frame(Base):
...
         self.items = []
         self.start = None
         self.end = None
 
-    def _finalize(self):
-        self.ports = tuple(self.ports)
-        self.items = tuple(self.items)
+    def _finalize(self) -> None:
+        self.ports = self.ports
+        self.items = self.items
```
Comment (hohav) [inline slippi/event.py:285]: We're converting to a tuple here as an optimization, since tuples use less memory than lists. This is also why the type hints were Sequence and not List.

Admittedly that should be explained in a comment, but please apply Chesterton's Fence (don't make a change without understanding the status quo first).

File: slippi/util.py [hunk 2]
```diff
@@ -49,20 +49,20 @@ def unpack(fmt, stream):
     return struct.unpack(fmt, bytes)
 
 
-def expect_bytes(expected_bytes, stream):
+def expect_bytes(expected_bytes: bytes, stream: BinaryIO) -> None:
     read_bytes = stream.read(len(expected_bytes))
     if read_bytes != expected_bytes:
-        raise Exception(f'expected {expected_bytes}, but got: {read_bytes}')
+        raise Exception(f'expected {expected_bytes.decode()}, but got: {read_bytes.decode()}')
```
Comment (hohav) [inline slippi/util.py:55]: This seems unrelated to the rest of the PR.

**Review by hohav — CHANGES_REQUESTED** [review]
Again, please open a discussion before beginning a large refactor. Also this should be broken into smaller PRs (one to add the type annotations, one to change how mypy is called, maybe more for any unrelated changes).

I'm ambivalent about the main change (adding type annotations everywhere) because it adds a lot of verbiage for unclear benefit. I intended to apply mypy in a targeted way, mostly to functions that are part of py-slippi's public API, because I think that provides 80% of the benefit.

On the other hand, type annotations do provide some benefit for internal APIs by acting as documentation. I'm not sure it's worth the verbosity, but perhaps my perspective is skewed as the author of the code. How helpful would you find these when reading py-slippi's code?

File: slippi/event.py (hunk 1 above)
Comment (DustinAlandzes) [reply slippi/event.py:285]: Sequence doesn't support indexing or .append because it's immutable
```
slippi/parse.py:159: error: Unsupported target for indexed assignment ("Sequence[Optional[Port]]")
slippi/parse.py:173: error: "Sequence[Item]" has no attribute "append"
```

https://justincaustin.com/blog/mimicking-immutability-python-type-hints/
>Now, what if we want to have an immutable list or immutable dictionary. Unfortunately, Python does not have a frozenset alternative for lists or dicts. We can instead reach into the typing module for “immutable” versions, namely Sequence and Mapping respectively.

> Calling Sequence and Mapping “immutable” may be slightly off. They do not enforce immutability; instead, they just do not declare methods that can mutate the object.

image

Comment (DustinAlandzes) [reply slippi/event.py:285]: It seems like tuples are more efficient than lists but they're also immutable, and that's fine but they're not being used that way

File: slippi/util.py (hunk 2 above)
Comment (DustinAlandzes) [reply slippi/util.py]: I got this error, I just changed this to use repr() instead of .decode(). adding !r to the end worked too but seems too obscure
```
slippi/util.py:55: error: On Python 3 '{}'.format(b'abc') produces "b'abc'", not 'abc'; use '{!r}'.format(b'abc') if this is desired behavior
```

File: slippi/event.py (hunk 1 above)
Comment (DustinAlandzes) [reply slippi/event.py:285]: I think type hints makes the code easier to read/understand, and catches mistakes

Thinking about the types also make implicit type conversion obvious like the list/tuple thing

Comment (hohav) [reply slippi/event.py:285]: `Sequence` does support indexing. Re:mutation, there is no reason I can think of why a consumer of py-slippi's public API would need to mutate the data we return, and if they did they could just create a new tuple. However as an implementation detail we need these collections to be mutable during construction, which is why I used `Sequence` instead of `Tuple` (since both tuples and lists are sequences).

But since you correctly point out that `Sequence` doesn't support mutation, that doesn't quite work. We could use a separate field during construction with type `List`, but AFAIK that requires an extra entry in `__slots__` and therefore increases memory usage even after parsing is over, which I would like to avoid.

So I think the best option in these cases is to use the type hint `collections.abc.Sequence[...]` and then cast to `MutableSequence` internally as needed.

Comment (DustinAlandzes) [reply slippi/event.py:285]: by indexing I meant "indexed assignment" like in the type error
